### DISCUSSION TITLE
React Native - AtomWithStorage - AsyncStorage

### DISCUSSION THREAD

**Discussion by summitWT** [pmndrs/jotai#2302] [Q&A] [answered]
Hi,

I'm trying to create a wrapper atom around `atomWithStorage` using `AsyncStorage` in React Native. Here's my implementation:

```ts
import {atomWithStorage, createJSONStorage} from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {add, formatISO} from 'date-fns';

async function getItem<T>(key: string): Promise<T> {
  const value = await AsyncStorage.getItem(key);
  const parsed = JSON.parse(value);
  return parsed;
}

async function setItem<T>(
  key: string,
  value: T,
  expireInHours = 12,
): Promise<void> {
  if (typeof value === 'string') {
    value = JSON.parse(value);
  }
  const expireAt = add(new Date(), {hours: expireInHours});
  const updatedValue = {...value, expireAt: formatISO(expireAt)};
  await AsyncStorage.setItem(key, JSON.stringify(updatedValue));
}

async function removeItem(key: string): Promise<void> {
 await AsyncStorage.removeItem(key);
}

export const atomWithAsyncStorage = <T>(key: string, initialValue: T) => {
  return atomWithStorage<T>(
    key,
    initialValue,
    createJSONStorage<T>(() => ({
      getItem,
      setItem,
      removeItem,
    })),
    {getOnInit: true},
  );
};
```

I've removed the `console.log`s to make this easier to read, but I can see that I'm getting and setting objects properly (at least when the `setItem` and `getItem` functions are being called. 

Here's how I'm using the wrapper: 

```ts
export const deviceAtom = atomWithAsyncStorage<Phone>(
  'device-atom',
  {} as Phone,
);



export const selectedDeviceAtom = atom(
  async get => {
    return await get(deviceAtom);
  },
  async (_get, set, newPhone: Phone) => {
    return await set(deviceAtom, newPhone);
  },
);
```

The major issue I'm having is that when I relaunch the app, my previous selection isn't being persisted. I did some digging and this seems to be a somewhat known issue, and adding the `getOnInit` property should have fixed this issue? 

For some more context, I tried this with `react-native-mmkv` and it worked just fine. However, I cannot use that package anymore due to some constraints with my project. The major difference I see between `mmkv` and `AsyncStorage` is that `mmkv` is synchronous and `AsyncStorage` is asynchronous. I feel like this is the root issue and I'm not sure what I'm doing wrong here.

**Comment by dai-shi** [comment]
If `getOnInit` option helped your case, it's probably a good solution.
Though, I wonder why it can't update without the option.
Looking at the code I wonder if you need to use createJSONStorage, because you already handle your `T` type. Not sure if it's related to the issue.

**Reply by summitWT** [reply, marked as answer]
Hi so I think I finally figured it out. Here's what I ended up doing:

```ts
import {atomWithStorage, createJSONStorage} from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {add, formatISO} from 'date-fns'

export const atomWithAsyncStorage = <T>(key: string, initialValue: T) => {
  const storage = createJSONStorage<T>(() => AsyncStorage);

  // preserve original getItem and setItem
  const {getItem, setItem} = storage;

  // override setItem
  storage.setItem = (_, value, expireInHours = 12) => {
    // add expireAt to newValue
    const expireAt = add(new Date(), {hours: expireInHours});
    const updatedValue = {...value, expireAt: formatISO(expireAt)};

    // updatedValue is a JSON object -- createJSONStorage handles that for us
    // call original setItem with updatedValue
    return setItem(key, updatedValue);
  };

  // override getItem
  storage.getItem = async () => {
    // call original getItem
    const value = await getItem(key, initialValue);

    // value is already a JSON object -- createJSONStorage handles that for us
    return value;
  };

  return atomWithStorage(key, initialValue, storage);
};
```

For some reason this pattern worked, but the previous one did not. Everything is persisting through app relaunches now 😃

**Reply by dai-shi** [reply]
Alternatively, you don't use createJSONStorage, and implement your own. I'd prefer it as it reads better with replacer/reviver options. Maybe we can add that to `createJSONString`.
